File: next/schedules.ts
/**
 * The schedule ticker and its REST surface: what actually makes a recurring run
 * happen (spec §9).
 *
 * THE INVARIANT THIS HOLDS: **a schedule that missed N slots fires ONCE.** The
 * arithmetic that guarantees it lives in `hostfn/schedule.ts` (`nextRun(spec, now)`
 * always advances from *now*); this file is the other half — the loop that reads
 * "due" as a boolean about one row rather than as a count of elapsed slots. Three
 * details carry it, and each is a bug if it slips:
 *
 *   1. `dueSchedules(now)` returns each enabled row **once**, however far behind it
 *      is. There is no inner loop walking a stale `next_run_at` forward slot by slot,
 *      because that loop is exactly the burst — a laptop shut overnight would wake up
 *      and open sixteen sessions before the user finished logging in.
 *   2. The advance happens **before** the fire, not after. A `fire` that throws must
 *      not leave the row still due: the very next tick, 30 seconds later, would fire
 *      it again, and again, for as long as the failure lasts.
 *   3. `now` is threaded in, never read inside. One tick uses one instant for both the
 *      due test and the advance, so a slow pass cannot skip a slot it should have
 *      caught, and a test can drive five hours of downtime in a millisecond.
 *
 * WHAT FIRING IS. A fresh **root** session titled from the schedule, with the
 * schedule's prompt posted into it as an ordinary user message. Not a branch, not a
 * subagent: the prompt has to stand alone anyway (the fired session sees no prior
 * conversation), and a root is the thing that shows up in the session list, forks,
 * and can be adopted — after the first turn starts it is indistinguishable from a
 * session the user opened by hand.
 *
 * WHY THE TICKER TAKES A CTX RATHER THAN A DATABASE. Firing needs the bus (to
 * announce the new session) and the turn starter (to run the prompt), and both arrive
 * on the ctx that `server/main.ts` builds. `startTurn` is read off it structurally,
 * the same seam `server/sessions.ts` and `agents/notes.ts` use — an unwired starter
 * degrades to "the session exists with an unanswered message", never to a crash.
 *
 * Ported from `src/schedules.ts` (`fireSchedule`, `tickSchedules`,
 * `startScheduleTicker`, and the REST handlers in `src/server/app.ts`). Deltas from
 * that port are marked `NOTE:`.
 */
import { CreateScheduleBody, PatchScheduleBody } from "./schema/requests.ts";
import type { Message, Schedule, Session } from "./schema/parts.ts";
import type { AppCtx, Db } from "./types.ts";
import {
  nextRun,
  scheduleCreate,
  type ScheduleDeps,
  schedulePatch,
  scheduleRemove,
} from "./hostfn/schedule.ts";
import { json, parseBody } from "./server/http.ts";
import type { WithTurnStarter } from "./server/sessions.ts";

// ---------------------------------------------------------------------------
// Firing
// ---------------------------------------------------------------------------

/** The production interval. ~30s, per spec §9. */
export const TICK_MS = 30_000;

export interface FireDeps extends ScheduleDeps {
  /** Where a failure to start the fired turn is reported. Tests pass a collector. */
  reportError?: (error: unknown, schedule: Schedule) => void;
}

/** What a firing produced, for the caller that wants to assert on it. */
export interface FiredSchedule {
  session: Session;
  message: Message;
}

/**
 * Fire one schedule: a fresh root session, the prompt as its first user message, and
 * a turn on it.
 *
 * The session is announced before the message, and the message before the turn, so a
 * live TUI renders the new session already carrying its prompt rather than an empty
 * card that fills in a beat later.
 *
 * **This never throws.** It is called from a timer callback with nobody to report to;
 * a throw would surface as an unhandled rejection and take the server down, losing
 * every other session with it. A failure to start the turn leaves the session and its
 * message persisted — the user can see what was supposed to run and post into it.
 */
export function fireSchedule(
  ctx: AppCtx,
  schedule: Schedule,
  deps: FireDeps = {},
): FiredSchedule | null {
  const now = deps.now ?? ctx.now ?? Date.now;
  const report = deps.reportError ??
    ((error: unknown, s: Schedule) =>
      console.error(`schedule ${s.id} (${s.title}) failed to fire:`, error));

  try {
    const session = ctx.db.createSession({
      id: crypto.randomUUID(),
      title: schedule.title,
      kind: "root",
      // A fired session inherits nothing: no parent thread, no lineage edge. The
      // prompt is the whole briefing, which is what the prompt section tells the
      // model when it writes one (`prompt/schedule.md`).
      parentId: null,
      createdAt: now(),
      ...(schedule.workspace
        ? { workspace: schedule.workspace, originDir: schedule.workspace }
        : {}),
    });
    ctx.bus.publish({ type: "session.created", sessionId: session.id, data: session });

    const message = ctx.db.createMessage({
      id: crypto.randomUUID(),
      sessionId: session.id,
      role: "user",
      parts: [{ type: "text", text: schedule.prompt }],
      pending: false,
      createdAt: now(),
    });
    indexQuietly(ctx.db, message);
    ctx.bus.publish({ type: "message.started", sessionId: session.id, data: message });

    // Fire and forget, like the HTTP post path: the turn runs for minutes and there
    // is no response to hold open. An absent starter is not an error — it is the
    // pre-M2 shape, and the session is still there with its prompt.
    const start = (ctx as AppCtx & WithTurnStarter).startTurn;
    if (start) {
      const running = start(ctx, session, message);
      if (running instanceof Promise) running.catch((err) => report(err, schedule));
    }
    return { session, message };
  } catch (err) {
    report(err, schedule);
    return null;
  }
}

/** Indexing failure is a degraded search, never a lost firing. */
function indexQuietly(db: Db, message: Message): void {
  try {
    db.indexMessage(message);
  } catch (err) {
    console.error(`failed to index message ${message.id}:`, err);
  }
}

// ---------------------------------------------------------------------------
// One tick
// ---------------------------------------------------------------------------

/**
 * One ticker pass at `now`. Returns the schedules that fired, in order.
 *
 * Read the two statements in the loop together — they are the catch-up rule:
 * `markScheduleRun` stamps `last_run_at = now` and `next_run_at = nextRun(spec, now)`
 * **before** `fire` runs, so the row is no longer due whatever happens next, and the
 * new time is measured from this instant rather than from the slot that was missed.
 *
 * `fire` is a parameter rather than a call to `fireSchedule` so a test can drive the
 * loop with a counter and prove the burst does not happen without a database full of
 * sessions or a single LLM call.
 */
export function tickSchedules(
  db: Db,
  now: number,
  fire: (schedule: Schedule) => void,
): Schedule[] {
  const due = db.dueSchedules(now);
  for (const schedule of due) {
    db.markScheduleRun(schedule.id, now, nextRun(schedule.spec, now));
    try {
      fire(schedule);
    } catch (err) {
      // `fireSchedule` does not throw; a test fake or a future caller might. One
      // schedule's failure must not skip the rest of this pass.
      console.error(`schedule ${schedule.id} (${schedule.title}) failed to fire:`, err);
    }
  }
  return due;
}

export interface TickerDeps extends FireDeps {
  /** Defaults to `TICK_MS`. */
  intervalMs?: number;
  /** Defaults to `fireSchedule`. Tests observe firings here. */
  fire?: (ctx: AppCtx, schedule: Schedule) => unknown;
}

/**
 * The production loop: `tickSchedules` on a ~30s interval. Returns a stopper.
 *
 * The timer is unref'd so the ticker alone never keeps a process alive — a CLI or a
 * test that starts one and finishes its work should exit, not hang for 30 seconds
 * waiting for a tick nobody reads.
 *
 * NOTE: no immediate pass at boot. The first tick lands one interval in, which gives
 * a server that is still recovering orphaned turns (`server/main.ts`) a moment before
 * it starts opening new sessions. A schedule due right now waits at most 30 seconds,
 * which is inside the cadence anything expressible in this grammar asked for.
 */
export function startScheduleTicker(ctx: AppCtx, deps: TickerDeps = {}): () => void {
  const now = deps.now ?? ctx.now ?? Date.now;
  const fire = deps.fire ?? ((c: AppCtx, s: Schedule) => fireSchedule(c, s, deps));
  const timer = setInterval(() => {
    try {
      tickSchedules(ctx.db, now(), (schedule) => fire(ctx, schedule));
    } catch (err) {
      // A throwing tick must not kill the interval — the next pass may well work,
      // and a silently dead ticker is a feature that stops existing with no signal.
      console.error("schedule tick failed:", err);
    }
  }, deps.intervalMs ?? TICK_MS);
  Deno.unrefTimer(timer);
  return () => clearInterval(timer);
}

// ---------------------------------------------------------------------------
// REST
// ---------------------------------------------------------------------------

/**
 * The four handlers are `function` DECLARATIONS, not `const` arrows, and that is
 * load-bearing rather than stylistic.
 *
 * This module and `server/app.ts` form an import cycle — app.ts imports these
 * handlers for its route table, and this file imports app.ts's `json`/`parseBody`.
 * A cycle is fine when every binding read during module evaluation is already
 * initialized, and hoisted function declarations are: they exist from module
 * *instantiation*, before any body runs. A `const` handler would be in its temporal
 * dead zone whenever this module happens to be evaluated first (a test importing
 * `schedules.ts` directly does exactly that), and app.ts's route table would throw
 * building itself. Same reason app.ts's own `json`, `parseBody` and `route` are
 * declarations.
 */

/**
 * `GET /schedules` — every schedule, in creation order.
 *
 * No visibility derivation here: schedules are flat and few, and the panel shows
 * disabled ones too — that is how you re-enable one.
 */
export function listSchedulesH(_req: Request, ctx: AppCtx): Response {
  return json(ctx.db.listSchedules());
}

/** `POST /schedules` — 201 with the stored row, `next_run_at` already computed. */
export async function createScheduleH(req: Request, ctx: AppCtx): Promise<Response> {
  const body = await parseBody(req, CreateScheduleBody);
  return json(await scheduleCreate(ctx.db, body, { now: ctx.now }), 201);
}

/** `PATCH /schedules/:id` — partial update; `workspace: null` clears it. */
export async function patchScheduleH(
  req: Request,
  ctx: AppCtx,
  params: Record<string, string>,
): Promise<Response> {
  // `{}` as the fallback: every field is optional, so an empty body is a legal
  // no-op patch rather than a 400 about a missing object.
  const body = await parseBody(req, PatchScheduleBody, {});
  return json(await schedulePatch(ctx.db, params.id, body, { now: ctx.now }));
}

/** `DELETE /schedules/:id` — 404 on an unknown id rather than a silent success. */
export function deleteScheduleH(
  _req: Request,
  ctx: AppCtx,
  params: Record<string, string>,
): Response {
  scheduleRemove(ctx.db, params.id);
  return json({ ok: true, removed: params.id });
}

File: bough-rs/crates/bough-core/src/schedules.rs
//! The schedule ticker + fire + report-back (port of `src/schedules.ts`) and
//! the spec grammar's pure math (port of `src/hostfn/schedule.ts`'s
//! `parseSpec`/`nextRun` — the CRUD half lands with `hostfn::schedule`).
//!
//! The invariant: **a schedule that missed N slots fires ONCE.** Three
//! load-bearing details: `due_schedules(now)` returns each enabled row once
//! (no catch-up loop); the advance happens BEFORE the fire (a throwing fire
//! must not leave the row due); `now` is threaded in, never read inside.
//! `fire_schedule` never panics (timer-callback context); the report-back
//! note's outcome is read from the DATABASE, not the settled future.
//!
//! The arithmetic half of the same invariant: **`next_run_at` is always
//! computed FROM NOW, never from the stale stored value.** A laptop closed
//! overnight with an `every:30m` schedule wakes up 16 slots behind; advancing
//! from `now` means one run, then the cadence resumes. [`next_run`] measures
//! from the instant it is handed, so the catch-up rule is provable as
//! arithmetic.
//!
//! v1 STUB (root.md §8): [`parse_spec`]/[`next_run`] ship here with wave 1;
//! the ticker, firing, report-back and REST land with wave 2 row 2.8 — until
//! then [`fire_schedule`] is wired to `None` and the ticker never starts.

use std::sync::LazyLock;

use chrono::{DateTime, Duration, Local, LocalResult, NaiveDateTime, TimeZone};
use regex::Regex;

use crate::errors::{BoughError, ErrorKind};
use crate::hostfn::schedule::{ParsedSpec, SPEC_HELP};
use crate::schema::parts::{Message, Schedule, Session};
use crate::types::{AppCtx, Db};

pub const TICK_MS: u64 = 30_000;

/// Stable marker text the creator's model and UI key off.
pub const SCHEDULE_NOTE_PREFIX: &str = "[schedule fired]";

// ---------------------------------------------------------------------------
// The grammar (pure)
// ---------------------------------------------------------------------------

static EVERY_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^every:(\d+)(m|h|d)$").unwrap());
static DAILY_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^daily@(\d{1,2}):(\d{2})$").unwrap());

/// Parse a spec string, or `None` when it does not match the grammar.
///
/// N ≥ 1: `every:0m` would parse to a zero interval, and a schedule whose next
/// run is always "now" fires on every single tick forever.
pub fn parse_spec(spec: &str) -> Option<ParsedSpec> {
    if let Some(caps) = EVERY_RE.captures(spec) {
        let n: i64 = caps[1].parse().ok()?;
        if n < 1 {
            return None;
        }
        let unit_ms: i64 = match &caps[2] {
            "m" => 60_000,
            "h" => 3_600_000,
            "d" => 86_400_000,
            _ => unreachable!("the regex admits only m|h|d"),
        };
        return Some(ParsedSpec::Every { ms: n.checked_mul(unit_ms)? });
    }
    if let Some(caps) = DAILY_RE.captures(spec) {
        let hh: u8 = caps[1].parse().ok()?;
        let mm: u8 = caps[2].parse().ok()?;
        if hh > 23 || mm > 59 {
            return None;
        }
        return Some(ParsedSpec::Daily { hh, mm });
    }
    None
}

/// Resolve a local wall-clock time to an instant, with the DST cases decided
/// EXPLICITLY (the TS relied on `Date.prototype.setHours`, which ECMA-262 pins
/// to the same two choices):
///
/// - **Ambiguous** (fall-back repeats the hour): the EARLIEST occurrence — the
///   pre-transition offset, the first time the clock shows HH:MM.
/// - **Nonexistent** (spring-forward skips the hour): the instant the
///   pre-transition offset names, which for the one-hour gap every US/EU zone
///   uses is the same instant as wall+1h under the new offset. "daily@02:30"
///   fires at 03:30 on the one morning 02:30 never happens.
fn resolve_local<Tz: TimeZone>(tz: &Tz, naive: NaiveDateTime) -> DateTime<Tz> {
    match tz.from_local_datetime(&naive) {
        LocalResult::Single(dt) => dt,
        LocalResult::Ambiguous(earliest, _latest) => earliest,
        LocalResult::None => match tz.from_local_datetime(&(naive + Duration::hours(1))) {
            LocalResult::Single(dt) => dt,
            LocalResult::Ambiguous(earliest, _latest) => earliest,
            // No real zone has two adjacent gaps; interpret as UTC rather
            // than panic in a timer path.
            LocalResult::None => tz.from_utc_datetime(&naive),
        },
    }
}

/// [`next_run_parsed`] against an explicit zone — the seam that makes the DST
/// cases testable deterministically (production uses [`Local`]).
pub fn next_run_in<Tz: TimeZone>(tz: &Tz, spec: ParsedSpec, from: i64) -> i64 {
    match spec {
        ParsedSpec::Every { ms } => from + ms,
        ParsedSpec::Daily { hh, mm } => {
            let local = tz
                .timestamp_millis_opt(from)
                .single()
                .expect("an instant maps to exactly one local time");
            let date = local.date_naive();
            let wall = |d: chrono::NaiveDate| {
                resolve_local(tz, d.and_hms_opt(hh as u32, mm as u32, 0)
                    .expect("parse_spec bounds hh/mm"))
            };
            let today = wall(date);
            // Strictly after, never equal: `next_run` is called at fire time
            // with the firing instant as `from`, and a result equal to `from`
            // would be due again on the very next tick.
            if today.timestamp_millis() > from {
                today.timestamp_millis()
            } else {
                wall(date.succ_opt().expect("not the end of the calendar")).timestamp_millis()
            }
        }
    }
}

/// The next fire time strictly after `from` (epoch ms), for an
/// already-parsed spec. `daily@` resolves in LOCAL wall-clock time — the run
/// stays at HH:MM local on either side of a DST transition, which is what a
/// user who asked for "every morning at nine" means.
pub fn next_run_parsed(spec: ParsedSpec, from: i64) -> i64 {
    next_run_in(&Local, spec, from)
}

/// The next fire time strictly after `from` (epoch ms). Errors with a 400
/// `ScheduleError` naming the grammar on a spec that does not parse.
pub fn next_run(spec: &str, from: i64) -> Result<i64, BoughError> {
    let parsed = parse_spec(spec).ok_or_else(|| {
        BoughError::http(
            400,
            ErrorKind::Schedule,
            format!("invalid schedule spec: {spec} — use {SPEC_HELP}"),
        )
    })?;
    Ok(next_run_parsed(parsed, from))
}

// ---------------------------------------------------------------------------
// Ticker + fire — v1 STUBS (wave 2, row 2.8)
// ---------------------------------------------------------------------------

/// What one firing produced: the fresh session and its prompt message.
#[derive(Clone, Debug)]
pub struct FiredSchedule {
    pub session: Session,
    pub message: Message,
}

/// Fire one schedule: fresh session + prompt message + turn. **Never
/// panics** (timer-callback context); `None` = the firing failed after being
/// reported.
///
/// v1 STUB: firing is a wave-2 behavior (row 2.8); until then no schedule
/// fires and this honestly reports "nothing fired".
pub fn fire_schedule(_ctx: &AppCtx, _schedule: &Schedule) -> Option<FiredSchedule> {
    None
}

/// One ticker pass over the due rows: advance each row FROM `now` first, then
/// hand it to `fire` — the advance-before-fire order is what makes a throwing
/// fire safe. Returns the due schedules in order; `fire` is a parameter for
/// testability.
///
/// v1 STUB: the ticker never runs in v1, so no row is ever due through this
/// path; the pass reports an empty tick.
pub fn tick_schedules(
    _db: &dyn Db,
    _now: i64,
    _fire: &mut dyn FnMut(&Schedule),
) -> Vec<Schedule> {
    Vec::new()
}

/// Start the interval loop; returns a stopper. No immediate pass at boot
/// (first tick lands one interval in — gives orphan-recovery a moment).
///
/// v1 STUB: no timer is started; the stopper is a no-op.
pub fn start_schedule_ticker(_ctx: &AppCtx) -> impl FnOnce() {
    || {}
}

// ---------------------------------------------------------------------------
// Tests — the pure math, ported from src/hostfn/schedule.test.ts
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Utc;

    /// `Date.UTC(2026, 0, 15, 12, 0, 0)`.
    fn t0() -> i64 {
        Utc.with_ymd_and_hms(2026, 1, 15, 12, 0, 0).unwrap().timestamp_millis()
    }

    const MINUTE: i64 = 60_000;
    const HOUR: i64 = 3_600_000;

    fn local_ms(y: i32, mo: u32, d: u32, h: u32, mi: u32) -> i64 {
        Local.with_ymd_and_hms(y, mo, d, h, mi, 0).single().unwrap().timestamp_millis()
    }

    #[test]
    fn parse_spec_accepts_every_n_m_h_d() {
        assert_eq!(parse_spec("every:30m"), Some(ParsedSpec::Every { ms: 30 * MINUTE }));
        assert_eq!(parse_spec("every:2h"), Some(ParsedSpec::Every { ms: 2 * HOUR }));
        assert_eq!(parse_spec("every:1d"), Some(ParsedSpec::Every { ms: 86_400_000 }));
    }

    #[test]
    fn parse_spec_accepts_daily_hh_mm() {
        assert_eq!(parse_spec("daily@09:00"), Some(ParsedSpec::Daily { hh: 9, mm: 0 }));
        assert_eq!(parse_spec("daily@9:05"), Some(ParsedSpec::Daily { hh: 9, mm: 5 }));
        assert_eq!(parse_spec("daily@23:59"), Some(ParsedSpec::Daily { hh: 23, mm: 59 }));
    }

    #[test]
    fn parse_spec_rejects_everything_else() {
        for bad in [
            "",
            "every:0m", // N ≥ 1 — a zero interval is always due, on every tick, forever
            "every:m",
            "every:5s",
            "every:5w",
            "every: 5m",
            "EVERY:5m",
            "daily@24:00",
            "daily@09:60",
            "daily@9",
            "daily@09:00:00",
            "0 9 * * *", // cron is NOT the grammar
            "hourly",
        ] {
            assert_eq!(parse_spec(bad), None, "expected {bad:?} to be rejected");
        }
    }

    #[test]
    fn next_run_for_every_adds_the_interval_to_the_instant_it_is_given() {
        assert_eq!(next_run("every:30m", t0()).unwrap(), t0() + 30 * MINUTE);
        // The invariant, stated as arithmetic: five hours of downtime does not
        // compound. Whatever `from` is, the answer is exactly one interval
        // later.
        assert_eq!(
            next_run("every:30m", t0() + 5 * HOUR).unwrap(),
            t0() + 5 * HOUR + 30 * MINUTE
        );
    }

    #[test]
    fn next_run_for_daily_lands_at_the_next_local_wall_clock_occurrence() {
        // Local time, so the assertion is built with the local constructor
        // rather than UTC.
        let morning = local_ms(2026, 1, 15, 8, 0);
        let nine = local_ms(2026, 1, 15, 9, 0);
        assert_eq!(next_run("daily@09:00", morning).unwrap(), nine);

        // Already past today → tomorrow, same wall clock.
        let afternoon = local_ms(2026, 1, 15, 14, 0);
        assert_eq!(next_run("daily@09:00", afternoon).unwrap(), local_ms(2026, 1, 16, 9, 0));

        // Exactly at the slot is NOT "now again": strictly after, or the row
        // stays due.
        assert_eq!(next_run("daily@09:00", nine).unwrap(), local_ms(2026, 1, 16, 9, 0));
    }

    #[test]
    fn next_run_errors_on_a_spec_that_does_not_parse() {
        let err = next_run("weekly", t0()).unwrap_err();
        assert_eq!(err.status(), 400);
        assert_eq!(err.name(), "ScheduleError");
        let message = err.to_string();
        assert!(message.contains("invalid schedule spec: weekly"), "message: {message}");
        // The grammar is in the message — the model's next move is to write
        // another spec, and "invalid" alone gets a second guess, not a fix.
        assert!(message.contains("every:<N><m|h|d>"), "message: {message}");
    }

    // ---- DST, pinned against a fixed zone (chrono-tz) ----------------------
    //
    // US DST in 2026: spring forward Sun Mar 8 02:00→03:00, fall back Sun
    // Nov 1 02:00→01:00 (America/Los_Angeles: PST = UTC-8, PDT = UTC-7).

    use chrono_tz::America::Los_Angeles as LA;

    fn utc_ms(y: i32, mo: u32, d: u32, h: u32, mi: u32) -> i64 {
        Utc.with_ymd_and_hms(y, mo, d, h, mi, 0).unwrap().timestamp_millis()
    }

    #[test]
    fn daily_stays_at_the_wall_clock_across_the_spring_forward_transition() {
        // From Sat Mar 7 10:00 PST (18:00Z), the next 09:00 is Sun Mar 8
        // 09:00 PDT (16:00Z) — 23 real hours later, same wall clock.
        let from = utc_ms(2026, 3, 7, 18, 0);
        assert_eq!(
            next_run_in(&LA, ParsedSpec::Daily { hh: 9, mm: 0 }, from),
            utc_ms(2026, 3, 8, 16, 0)
        );
    }

    #[test]
    fn daily_stays_at_the_wall_clock_across_the_fall_back_transition() {
        // From Sat Oct 31 10:00 PDT (17:00Z), the next 09:00 is Sun Nov 1
        // 09:00 PST (17:00Z) — 25 real hours later, same wall clock.
        let from = utc_ms(2026, 10, 31, 17, 0);
        assert_eq!(
            next_run_in(&LA, ParsedSpec::Daily { hh: 9, mm: 0 }, from),
            utc_ms(2026, 11, 1, 17, 0)
        );
    }

    #[test]
    fn a_nonexistent_wall_time_resolves_forward_across_the_gap() {
        // 02:30 never happens on Mar 8 2026 in LA; the run lands on the
        // instant the pre-transition offset names — 03:30 PDT (10:30Z).
        let from = utc_ms(2026, 3, 8, 8, 0); // Mar 8 00:00 PST
        assert_eq!(
            next_run_in(&LA, ParsedSpec::Daily { hh: 2, mm: 30 }, from),
            utc_ms(2026, 3, 8, 10, 30)
        );
    }

    #[test]
    fn an_ambiguous_wall_time_resolves_to_the_earliest_occurrence() {
        // 01:30 happens twice on Nov 1 2026 in LA; the run is the FIRST one —
        // 01:30 PDT (08:30Z), not 01:30 PST (09:30Z).
        let from = utc_ms(2026, 11, 1, 7, 0); // Nov 1 00:00 PDT
        assert_eq!(
            next_run_in(&LA, ParsedSpec::Daily { hh: 1, mm: 30 }, from),
            utc_ms(2026, 11, 1, 8, 30)
        );
    }

}
